### ISSUE TITLE
label.shape and euclidean_distance.shape are inconsistent

### ISSUE THREAD

**Issue by yanxinpan** [harveyslash/Facial-Similarity-with-Siamese-Networks-in-Pytorch#26]
![image](https://user-images.githubusercontent.com/10967302/55661890-afbab980-57c3-11e9-8f7c-8aea366a9eb2.png)
The loss is wrong due to shape inconsistent.

**Comment by harveyslash** [comment]
The loss and distance are two different things and they dont have to have the same shape. 

The distance tells us how far two images are in terms of similarity. 
The labels tell us wether or not two images are the same

**Comment by yanxinpan** [comment]
No, you didn't get my point here. To be more concrete:

The batch size is 50, meaning there are 50 pairs of input images. Each pair should have a distance telling us how far two images are in terms of similarity and a label telling us whether or not two images are the same. In addition, a loss value to tell us how good the distance is, so the loss shape should be [50] as well.

The only thing wrong here is: in your implementation, the label shape is [50,1], while distance shape is [50]. You multiply a 2D tensor by a 1D tensor, so you end up with a loss tensor with shape [50,50], which is a wrong shape and wrong value. 

If simply correct this by adding a line: label = label.unsqueeze(-1), then the tensor multiplication is correct and the siamese net has a better performance.

**Comment by harveyslash** [comment]
Label similarity values are not given during training time. If it was, this would be more of a regression problem.

**Comment by yanxinpan** [comment]
The code is wrong. Please check how pytorch handle 2D tensor and 1D tensor multiplication. 

Tensor A (with shape [50,1]) multiplying tensor B (with shape [50]) gets a tensor C (with shape [50,50])
Tensor A* (A* = A.squeeze(-1), A.shape: [50])  multiplying tensor B (with shape [50]) gets a tensor D (with shape [50])

A and A* have the same element-wise value, but the multiplication results are totally different.

do label=label.squeeze(-1) (there was a typo in my previous comments).
